GAN/main.swift

```diff
@@ -0,0 +1,213 @@
+let epochCount = 10
+let batchSize = 32
+let outputFolder = "./output/"
+let imageHeight = 28, imageWidth = 28
+let imageSize = imageHeight * imageWidth
+let latentSize = 64
+
+func plot(image: Tensor<Float>, name: String) {
+    // Create figure.
+    let ax = plt.gca()
+    let array = np.array([image.scalars])
+    let pixels = array.reshape(image.shape)
+    if !FileManager.default.fileExists(atPath: outputFolder) {
+        try! FileManager.default.createDirectory(atPath: outputFolder,
+                            withIntermediateDirectories: false,
+                                             attributes: nil)
+    }
+    ax.imshow(pixels, cmap: "gray")
+    plt.savefig("\(outputFolder)\(name).png", dpi: 300)
+    plt.close()
+}
+
+/// Reads a file into an array of bytes.
+func readFile(_ filename: String) -> [UInt8] {
+    let possibleFolders = [".", "Resources", "GAN/Resources"]
+    for folder in possibleFolders {
+        let parent = URL(fileURLWithPath: folder)
+        let filePath = parent.appendingPathComponent(filename).path
+        guard FileManager.default.fileExists(atPath: filePath) else {
+            continue
+        }
+        let d = Python.open(filePath, "rb").read()
+        return Array(numpy: np.frombuffer(d, dtype: np.uint8))!
+    }
+    print("Failed to find file with name \(filename) in the following folders: \(possibleFolders).")
+    exit(-1)
+}
+
+/// Reads MNIST images from specified file path.
+func readMNIST(imagesFile: String) -> Tensor<Float> {
+    print("Reading data.")
+    let images = readFile(imagesFile).dropFirst(16).map { Float($0) }
+    let rowCount = images.count / imageSize
+
+    print("Constructing data tensors.")
+    return Tensor(shape: [rowCount, imageHeight * imageWidth], scalars: images) / 255.0 * 2 - 1
+}
+
+// Models
+
+struct Generator: Layer {
+    var dense1 = Dense<Float>(inputSize: latentSize, outputSize: latentSize*2,
+                              activation: { leakyRelu($0) })
+    var dense2 = Dense<Float>(inputSize: latentSize*2, outputSize: latentSize*4,
+                              activation: { leakyRelu($0) })
+    var dense3 = Dense<Float>(inputSize: latentSize*4, outputSize: latentSize*8,
+                              activation: { leakyRelu($0) })
+    var dense4 = Dense<Float>(inputSize: latentSize*8, outputSize: imageSize,
+                              activation: tanh)
+    
+    var batchnorm1 = BatchNorm<Float>(featureCount: latentSize*2)
+    var batchnorm2 = BatchNorm<Float>(featureCount: latentSize*4)
+    var batchnorm3 = BatchNorm<Float>(featureCount: latentSize*8)
+    
+    @differentiable
+    func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        let x1 = batchnorm1(dense1(input))
+        let x2 = batchnorm2(dense2(x1))
+        let x3 = batchnorm3(dense3(x2))
+        return dense4(x3)
+    }
+}
+
+struct Discriminator: Layer {
+    var dense1 = Dense<Float>(inputSize: imageSize, outputSize: 256, activation: { leakyRelu($0) })
```

Comment:
```suggestion
    var dense1 = Dense<Float>(inputSize: imageSize, outputSize: 256,
                              activation: { leakyRelu($0) })
```
For consistency with other dense layers below.